File: Project Euler/Problem 18/maximum_path_sum_1.rb
# usage: ruby maximum_path_sum_1.rb <input file>

start_time = Time.new

triangle = File.read(ARGV[0]).to_str
max_level = triangle.count("\n")
triangle = triangle.split(/ |\n/)

triangle.each do |x|
	x.to_i
end

#puts triangle

def max_local(a, b)
	if a > b
		return a
	else
		return b
	end
end

puts max_level

def max_path_sum(triangle, level, index, max_level)

	if level > max_level
		return 0
	end 
	
	max_sum = 0;
	new_level = level + 1;

	rest =   	max_local(max_path_sum(triangle, new_level, index, max_level),
							max_path_sum(triangle, new_level, index + 1, max_level))	

	max_sum = triangle[index -1 + (level - 1) * level / 2].to_i + rest
	
	return max_sum;
end

puts max_path_sum(triangle, 1, 1, max_level)

puts "Time taken in ms: #{(Time.new - start_time) * 1000}"

File: Project Euler/Problem 18/maximum_path_sum_1.js
function max_path_sum_recurse(triangle, level, index, max_level) {
	
	if (level > max_level) return 0;
	
	var max_sum = 0;
	
	max_sum = triangle[index -1 + (level - 1) * level / 2] +
	 Math.max(max_path_sum_recurse(triangle, level + 1, index, max_level),
				max_path_sum_recurse(triangle, level + 1, index + 1, max_level));	

	//console.log(max_sum);
	return max_sum;
}


// usage:  node maximum_path_sum_1.js <input file>

var max_path_sum = function() {

	var fs = require("fs");
	
	var triangle = fs.readFileSync(process.argv[2]).toString();
	
	var level = 0;
	for (var i = 0; i < triangle.length; ++i)
	{
			if (triangle[i] == "\n") level++;
	}
	
	console.log(level);
	
	triangle = triangle.split(/ |\n/);
	triangle.pop();
	triangle = triangle.map( function(x){
		return parseInt(x, 10);
	});
	
	console.log(triangle);
	
	return max_path_sum_recurse(triangle, 1, 1, level);
}

console.log(max_path_sum());
